keps/sig-windows/4802-windows-node-shutdown/README.md

```diff
@@ -0,0 +1,889 @@
+*   Provide guarantee to handle all cases of graceful node shutdown, for
+    example abrupt shutdown or sudden power cable pull can’t result in graceful
+    shutdown
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+*   As a cluster administrator, I can configure the nodes in my cluster to
+    allocate X seconds for my pods to terminate gracefully during a node
+    shutdown
+
+#### Story 2
+
+*   As a developer I can expect that my pods will terminate gracefully during
+    node shutdowns
+
+### Background on Windows Shutdown
+
+In the context of this KEP, shutdown is referred to as shutdown or restart of the
+underlying Windows machine. A shutdown can be initiated via a
+variety of methods for example:
+
+1. `shutdown /t 0`
+2. `shutdown /t 30` `#schedule a delayed shutdown in 30 seconds`
+3. `shutdown /r /t 30` `#schedule a delayed restart in 30 seconds`
+4. Physically pressing the power button on the machine
+5. If a machine is a VM, the underlying hypervisor can press the “virtual”
+   power button
+6. For a cloud instance, stopping the instance via Cloud API, e.g. via `az vm stop -g xx -n xx`.
+   Depending on the cloud provider, this may result in virtual power button press by the underlying hypervisor
+7. Windows update (ex: Patch Tuesday) can also trigger a Windows node to be shutdown and rebooted
+
+In the context of kubernetes on Windows node, system is not
+aware of the pods and containers running on the machine and will simply
+kill them as regular Windows processes, that's the reason that we will need
+to monitor the system for shutdown events and behave accordingly.
+
+The high-level process of detecting a system shutdown on Windows is the same for all types of Windows applications and services. To detect a shutdown, 
+we create a callback function and register it in the system. When a certain Windows event occurs, the system calls the callback function, transferring 
+information about the event via input parameters. The callback function then analyzes the input parameter data, and executes code accordingly. In the 
+case of this feature related, the call back can detect the related (pre)shutdown event, and kill the pods gracefully before exiting
+
+
+There’s no one way to make all types of applications (console, GUI and Windows services) detect a node shutdown, different types of applications differ 
+in their syntax and functionality.
```

Comment:
Is there a reason for this statement? It feels irrelevant from a kubelet perspective as it does not manage anything other than Pods on the node.

Reply:
This is basically trying to explain why we chose to run kubelet itset as a windows service, the following paragraphs have more details on this. by applications I mean Console/GUI vs window services, not pods vs non-pods, hope this clarifies. thanks!